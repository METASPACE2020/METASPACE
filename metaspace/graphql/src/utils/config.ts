import * as config from 'config';
import { IConfig } from 'config';

export interface ImageCategoryConfig {
  type: string,
  path: string,
  storage_types: string[],
}

export interface Config {
  port: number;
  ws_port: number;
  img_storage_port: number;
  log: {
    level: string;
  };
  defaults: {
    adducts: string[];
    moldb_names: string[];
  };
  img_upload: {
    iso_img_fs_path: string;
  };
  categories: {
    iso_image: ImageCategoryConfig;
    optical_image: ImageCategoryConfig;
    raw_optical_image: ImageCategoryConfig;
  };
  services: {
    moldb_service_host: string;
    sm_engine_api_host: string;
  };
  db: {
    host: string;
    database: string;
    user: string;
    password: string;
  };
  elasticsearch: {
    index: string;
    host: string;
    port: number;
  };
  rabbitmq: {
    host: string;
    user: string;
    password: string;
  };
  redis: {
    host: string;
    port: string;
  };
  cookie: {
    secret: string;
  };
  google: {
    client_id: string;
    client_secret: string;
    callback_url: string;
  };
  web_public_url: string;
  slack: {
    webhook_url: string;
    channel: string;
  };
  jwt: {
    secret: string;
  };
  features: {
    graphqlMocks: boolean;
    impersonation: boolean;
  };
  aws:  {
    aws_access_key_id: string;
    aws_secret_access_key: string;
    aws_region: string;
  };
  metadataLookups: {
    colocalizationAlgos: Record<string, string>;
    defaultColocalizationAlgo: string;
  };
}

const _config = config as any as (Config & IConfig);

export default _config;
