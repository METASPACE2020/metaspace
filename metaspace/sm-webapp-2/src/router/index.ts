import { createRouter, createWebHistory } from 'vue-router'
import { Component } from 'vue'
import NotFound from '../modules/App/NotFoundPage.vue'
import AboutPage from '../modules/App/AboutPage'
import DatasetsPage from '../modules/Datasets/DatasetsPage.vue'

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ '../modules/Annotations/AnnotationsPage.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ '../modules/MetadataEditor/MetadataEditPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ '../modules/MetadataEditor/UploadPage.vue'),
  DatasetOverviewPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ '../modules/Datasets/overview/DatasetOverviewPage'),
  DatasetComparisonPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ '../modules/Datasets/comparison/DatasetComparisonPage'),
  DatasetBrowserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ '../modules/Datasets/imzml/DatasetBrowserPage'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/HelpPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Group/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Project/ProjectsListPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Group/GroupsListPage'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/PrivacyPage.vue'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ '../modules/Group/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ '../modules/Project/ViewProjectPage.vue'),

}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, Component>

export const routes : any =[
  { path: '/', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
  { path: '/about', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
  { path: '/annotations', name: 'annotations', component: asyncPages.AnnotationsPage },
  {
    path: '/datasets',
    component: DatasetsPage,
    children: [
      { path: '', component: asyncPages.DatasetTable },
      { path: 'summary', component: asyncPages.DatasetTable },
    ],
  },
  {
    path: '/datasets/:dataset_id/comparison',
    name: 'datasets-comparison',
    component: asyncPages.DatasetComparisonPage,
  },
  { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
  { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.DatasetTable },
  { path: '/dataset/:dataset_id', name: 'dataset-overview', component: asyncPages.DatasetOverviewPage },
  { path: '/dataset/:dataset_id/annotations', name: 'dataset-annotations', component: asyncPages.AnnotationsPage },
  { path: '/dataset/:dataset_id/browser', name: 'dataset-browser', component: asyncPages.DatasetBrowserPage },
  { path: '/dataset/:dataset_id/enrichment', name: 'dataset-enrichment', component: asyncPages.DatasetTable },
  { path: '/upload', component: asyncPages.UploadPage },
  { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },



  { path: '/groups', component: asyncPages.GroupsListPage },
  { path: '/group/create', component: asyncPages.CreateGroupPage },
  { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },


  { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
  { // Legacy URL sent in "request access" emails up until Feb 2019
    path: '/project/:projectIdOrSlug/manage',
    redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } },
  },
  { path: '/projects', component: asyncPages.ProjectsListPage },



  { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },


  {
    path: '/:pathMatch(.*)*',
    component: NotFound,
  },
]

const router = createRouter({
  // @ts-ignore
  history: createWebHistory(import.meta.env.BASE_URL),
  routes,
  scrollBehavior(to, from, savedPosition) {
    // return desired position
    if (savedPosition) {
      return savedPosition;
    }
    return { top: 0 };
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if (to.path !== from.path && daysSincePageLoad > 1.75 || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})


const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
