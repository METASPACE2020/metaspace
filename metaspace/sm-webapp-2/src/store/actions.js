import apolloClient from '../api/graphqlClient';
import {fetchOptionListsQuery} from '../api/metadata';
import {decodeParams} from '../modules/Filters';
import {computed} from "vue";
import store from "../store/index";
import reportError from "@/lib/reportError";
export default {

  async initFilterLists(context) {
    if(context.state.filterListsLoading || context.state.filterLists != null)
      return;

    context.commit('setFilterListsLoading');
    let response = {}

    try{
      response = await apolloClient.query({
        query: fetchOptionListsQuery,
        fetchPolicy: 'cache-first',
      });
    } catch (err) {
      reportError(err);
    }

    // set annotationIds default according to values passed from snapshot to the store state
    context.commit('setFilterLists', {...response.data,
      annotationIds: computed(() => store.state.snapshotAnnotationIds)} );

    // Refresh the current filter so that computed defaults that depend on `filterLists` are applied
    const filter = decodeParams(context.state.route, context.state.filterLists)
    context.commit('updateFilter', filter)
  },

};
