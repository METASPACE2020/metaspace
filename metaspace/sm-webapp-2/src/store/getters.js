export default {
  filter() {
    return {}
    // return decodeParams(state.route, state.filterLists);
  },

  currentUser(state) {
    return state.currentUser;
  },
}
