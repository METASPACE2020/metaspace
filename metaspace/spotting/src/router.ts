/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter from 'vue-router'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  DashboardPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Dashboard/DashboardPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(dashboard)?', name: 'dashboard', component: asyncPages.DashboardPage },
    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if (to.path !== from.path && daysSincePageLoad > 1.75 || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
