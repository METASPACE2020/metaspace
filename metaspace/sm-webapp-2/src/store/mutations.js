import { isEqual, omit, without } from 'lodash-es'
import router from '../router'
import compare from '../lib/compare'
import {
  decodeParams,
  encodeParams,
  encodeSections,
  encodeSortOrder,
  FILTER_SPECIFICATIONS,
  getFilterInitialValue,
  stripFilteringParams,
} from '../modules/Filters'
import { DEFAULT_ANNOTATION_VIEW_SECTIONS, DEFAULT_COLORMAP, DEFAULT_TABLE_ORDER } from '../modules/Filters/url'
import { DEFAULT_SCALE_TYPE } from '../lib/constants'

function updatedLocation(state, filter) {
  const query = encodeParams(filter, state.route.path, state.filterLists)

  state.lastUsedFilters[state.route.path] = {
    filter,
    query,
    order: state.orderedActiveFilters,
  }

  return {
    query: Object.assign(query, stripFilteringParams(state.route.query)),
  }
}

function routerReplace(state, params) {
  router.replace(params)
  state.route.path = params.path || state.route.path
  state.route.params = params.params || state.route.params
  state.route.query = params.query || state.route.query
}

function routerPush(state, params) {
  router.push(params)
  state.route.path = params.path || state.route.path
  state.route.params = params.params || state.route.params
  state.route.query = params.query || state.route.query
}

function replaceURL(state, filter) {
  routerReplace(state, updatedLocation(state, filter))
}

function pushURL(state, filter) {
  routerPush(state, updatedLocation(state, filter))
}

function sortFilterKeys(keys) {
  keys.sort((a, b) => compare(FILTER_SPECIFICATIONS[a].sortOrder || 100, FILTER_SPECIFICATIONS[b].sortOrder || 100))
}

function updateFilter(state, filter, routerAction = null) {
  const active = []

  // drop unset filters
  for (let i = 0; i < state.orderedActiveFilters.length; i++) {
    const key = state.orderedActiveFilters[i]
    if (filter[key] !== undefined) active.push(key)
  }

  // append newly added filters to the end
  for (const key in filter) if (filter[key] !== undefined && active.indexOf(key) == -1) active.push(key)

  // sort
  sortFilterKeys(active)

  const changedFilterSet = !isEqual(state.orderedActiveFilters, active)

  state.orderedActiveFilters = active
  if (routerAction === 'push' || (routerAction == null && changedFilterSet)) pushURL(state, filter)
  else replaceURL(state, filter)
}

export default {
  updateRoute(state, route) {
    state.route.path = route.path
    state.route.params = route.params
    state.route.query = route.query
  },

  updateFilter(state, filter) {
    // TODO: Stop using updateFilter for adding/removing filters, as it makes reacting to added/removed filters
    // much more difficult
    updateFilter(state, filter)
  },

  updateCurrentUser(state, user) {
    state.currentUser = user
  },

  setReportError(state) {
    state.reportError = !state.reportError
  },

  replaceFilter(state, filter) {
    updateFilter(state, filter, 'replace')
  },

  addFilter(state, name) {
    const oldFilter = decodeParams(state.route, state.filterLists)
    const filtersToAdd = [name]
    const filtersToRemove = []

    // Check for required additional filters & conflicting filters
    ;(FILTER_SPECIFICATIONS[name].dependsOnFilters || []).forEach((key) => {
      if (oldFilter[key] === undefined) {
        filtersToAdd.push(key)
      }
    })
    ;(FILTER_SPECIFICATIONS[name].conflictsWithFilters || []).forEach((key) => {
      if (oldFilter[key] !== undefined) {
        filtersToRemove.push(key)
      }
    })

    const newFilter = { ...oldFilter }
    filtersToAdd.forEach((key) => {
      newFilter[key] = getFilterInitialValue(key, state.filterLists)
    })
    filtersToRemove.forEach((key) => {
      newFilter[key] = undefined
    })
    const newActive = without([...state.orderedActiveFilters, ...filtersToAdd], ...filtersToRemove)
    sortFilterKeys(newActive)

    state.orderedActiveFilters = newActive
    pushURL(state, newFilter)
  },

  removeFilter(state, name) {
    const oldFilter = decodeParams(state.route, state.filterLists)
    const filtersToRemove = [name]

    // Check for dependent filters that should also be removed
    Object.keys(oldFilter).forEach((key) => {
      const { dependsOnFilters } = FILTER_SPECIFICATIONS[key]
      if (dependsOnFilters != null && dependsOnFilters.includes(name)) {
        filtersToRemove.push(key)
      }
    })

    const newFilter = omit(oldFilter, filtersToRemove)
    state.orderedActiveFilters = without(state.orderedActiveFilters, ...filtersToRemove)
    pushURL(state, newFilter)
  },

  updateFilterOnNavigate(state, to) {
    const newActiveFilters = Object.keys(decodeParams(to, state.filterLists))

    const removed = without(state.orderedActiveFilters, ...newActiveFilters)
    const added = without(newActiveFilters, ...state.orderedActiveFilters)

    state.orderedActiveFilters = without(state.orderedActiveFilters, ...removed)
    state.orderedActiveFilters.push(...added)
    sortFilterKeys(state.orderedActiveFilters)
  },

  setFilterListsLoading(state) {
    state.filterListsLoading = true
  },

  setFilterLists(state, filterLists) {
    state.filterLists = filterLists
    state.filterListsLoading = false
  },

  setAnnotation(state, annotation) {
    state.annotation = annotation
  },

  setNormalizationMatrix(state, normalizationMatrix) {
    state.normalization = normalizationMatrix
  },

  resetChannels(state) {
    state.channels = []
  },

  addChannel(state, { id, annotations, settings }) {
    state.channels.push({ id, annotations, settings })
  },

  restoreChannels(state, channels) {
    state.channels = channels
  },

  removeChannel(state, { index }) {
    const aux = state.channels.slice(0)
    aux.splice(index, 1)
    state.channels = aux
  },

  updateChannel(state, { index, id, annotations, settings }) {
    const aux = state.channels.slice(0)
    aux[index] = { id, annotations, settings }
    state.channels = aux
  },

  setViewerMode(state, mode = 'SINGLE') {
    state.mode = mode
  },

  setRoiInfo(state, { key, roi }) {
    state.roiInfo[key] = roi
  },

  toggleRoiVisibility(state, visible) {
    state.roiInfo['visible'] = visible
  },

  resetRoiInfo(state) {
    state.roiInfo = { visible: false }
  },

  setSnapshotAnnotationIds(state, annotation) {
    state.snapshotAnnotationIds = annotation
  },

  updateAnnotationTableStatus(state, isLoading) {
    state.tableIsLoading = isLoading
  },

  startTour(state, tourDescription) {
    state.currentTour = tourDescription
  },

  endTour(state) {
    state.currentTour = null
  },

  updateAnnotationViewSections(state, activeSections) {
    const sections = encodeSections(activeSections)
    const defaultSections = encodeSections(DEFAULT_ANNOTATION_VIEW_SECTIONS)
    routerReplace(state, {
      query: sections !== defaultSections ? { ...state.route.query, sections } : omit(state.route.query, 'sections'),
    })
  },

  setColormap(state, cmap) {
    routerReplace(state, {
      query: cmap !== DEFAULT_COLORMAP ? { ...state.route.query, cmap } : omit(state.route.query, 'cmap'),
    })
  },

  setScaleType(state, scaleType) {
    routerReplace(state, {
      query:
        scaleType !== DEFAULT_SCALE_TYPE
          ? { ...state.route.query, scale: scaleType }
          : omit(state.route.query, 'scale'),
    })
  },

  setLockTemplate(state, lockTemplate) {
    routerReplace(state, {
      query: lockTemplate ? { ...state.route.query, lock: lockTemplate } : omit(state.route.query, 'lock'),
    })
  },

  setNormalization(state, normalization) {
    routerReplace(state, {
      query: !normalization ? omit(state.route.query, 'norm') : { ...state.route.query, norm: normalization },
    })
  },

  setCurrentPage(state, page) {
    routerReplace(state, {
      query: page !== 1 ? { ...state.route.query, page: String(page) } : omit(state.route.query, 'page'),
    })
  },

  setRow(state, row) {
    routerReplace(state, {
      query: row !== 1 ? { ...state.route.query, row: String(row) } : omit(state.route.query, 'row'),
    })
  },

  setSortOrder(state, sortOrder) {
    const sort = encodeSortOrder(sortOrder)
    const defaultSort = encodeSortOrder(DEFAULT_TABLE_ORDER)
    routerReplace(state, {
      query: sort !== defaultSort ? { ...state.route.query, sort } : omit(state.route.query, 'sort'),
    })
  },

  setColocalizationAlgo(state, colocalizationAlgo) {
    routerReplace(state, {
      query:
        colocalizationAlgo != null ? { ...state.route.query, alg: colocalizationAlgo } : omit(state.route.query, 'alg'),
    })
  },

  setDatasetTab(state, tab) {
    const path = {
      list: '/datasets',
      summary: '/datasets/summary',
    }[tab.toLowerCase()]
    routerReplace(state, {
      path,
      query: state.route.query,
    })
  },
}
