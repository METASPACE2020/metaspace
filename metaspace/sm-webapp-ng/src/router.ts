import { createRouter, createWebHistory } from 'vue-router'
import HomeView from './views/HomeView.vue'
import AboutView from './views/AboutView.vue'
import store from './store'
// ... other imports ...

const routes = [
  {
    path: '/',
    component: HomeView
  },
  {
    path: '/user/:id',
    component: AboutView
  }
]

const router = createRouter({
  history: createWebHistory(),
  routes
})

router.afterEach((to) => {
  console.log('to', to)
  // store.commit('updateFilterOnNavigate', to)
})

export default router
