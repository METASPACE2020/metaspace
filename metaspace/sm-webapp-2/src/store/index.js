import { createStore } from 'vuex'
import mutations from './mutations'
import getters from './getters'
import actions from './actions.js'

import { accountModule } from '../modules/Account'

const store = createStore({
  state: {
    // names of currently shown filters
    orderedActiveFilters: [],

    reportError: false,

    filterLists: null,
    filterListsLoading: false,

    // currently selected annotation
    annotation: undefined,

    // currently selected normalization matrix
    normalization: undefined,

    // roi settings
    roiInfo: { visible: false },

    // is annotation table loading?
    tableIsLoading: true,

    lastUsedFilters: {},

    currentTour: null,

    currentUser: {},

    // ion image global viewer settings
    channels: [],
    mode: 'SINGLE',
    route: {
      path: '',
      params: {},
      query: {},
    },
  },

  getters,
  mutations,
  actions,
  modules: {
    account: accountModule,
  },
})

export default store
