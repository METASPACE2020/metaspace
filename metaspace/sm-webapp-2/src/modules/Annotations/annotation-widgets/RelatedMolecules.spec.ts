import { nextTick, ref } from 'vue'
import router from '../../../router'
import store from '../../../store'
import { flushPromises, mount } from '@vue/test-utils'
import RelatedMolecules from './RelatedMolecules.vue'
import config from '../../../lib/config'

const testCompound = (name: string, id: string) => ({
  name,
  imageURL: `/mol-images/HMDB/${id}.svg`,
  information: [
    {
      databaseId: id,
      database: 'HMDB-v4',
      url: `http://www.hmdb.ca/metabolites/${id}`,
    },
  ],
})

const referenceAnnotation = {
  id: '2019-02-12_15h55m06s_HMDB-v4_2018-04-09_C10H11NO_plus_Na',
  ion: 'C10H11NO+Na+',
  fdrLevel: 0.1,
  dataset: { id: '2019-02-12_15h55m06s' },
  possibleCompounds: [testCompound('Tryptophanol', 'HMDB0003447')],
}
const relatedAnnotation = {
  ...referenceAnnotation,
  id: '2019-02-12_15h55m06s_HMDB-v4_2018-04-09_C10H13NO2_minus_H2O_plus_Na',
  ion: 'C10H13NO2-H2O+Na+',
  fdrLevel: 0.2,
  possibleCompounds: [testCompound('Maltoxazine', 'HMDB0030372'), testCompound('Salsolinol', 'HMDB0042012')],
}

vi.mock('@vue/apollo-composable', () => ({
  useQuery: vi.fn(() => ({
    result: ref({ allAnnotations: [relatedAnnotation, referenceAnnotation] }),
    loading: ref(false),
    error: ref(null),
    subscribeToMore: vi.fn(),
  })),
}))

describe('RelatedMolecules', () => {
  const propsData = { annotation: referenceAnnotation, databaseId: 22 }

  it('should match snapshot', async () => {
    config.features.isomers = true
    config.features.isobars = false
    const wrapper = mount(RelatedMolecules, {
      propsData,
      global: {
        plugins: [router, store],
      },
    })

    await flushPromises()
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()

    expect(wrapper.findAll('.ion-link').map((el) => el.text())).toEqual([
      expect.stringContaining('C10H11NO'),
      expect.stringContaining('C10H13NO2'),
    ])

    expect(wrapper.findAll('.compound').map((el) => el.text())).toEqual([
      expect.stringContaining('Tryptophanol'),
      expect.stringContaining('Maltoxazine'),
      expect.stringContaining('Salsolinol'),
    ])
  })

  it('should hide the isomer when the feature flag is off', async () => {
    config.features.isomers = false
    config.features.isobars = false
    const wrapper = mount(RelatedMolecules, {
      propsData,
      global: {
        plugins: [router, store],
      },
    })

    await flushPromises()
    await nextTick()

    expect(wrapper.find('.ion-link').exists()).toBe(false)
    expect(wrapper.findAll('.compound').map((el) => el.text())).toEqual([expect.stringContaining('Tryptophanol')])
  })
})
