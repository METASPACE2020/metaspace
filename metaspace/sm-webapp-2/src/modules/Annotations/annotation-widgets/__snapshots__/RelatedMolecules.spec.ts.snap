// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RelatedMolecules > should match snapshot 1`] = `
"<div data-v-bd75ece3=\\"\\" class=\\"related-molecules\\">
  <div data-v-bd75ece3=\\"\\">
    <div data-v-bd75ece3=\\"\\" class=\\"el-divider el-divider--horizontal\\" style=\\"--el-border-style: solid;\\" role=\\"separator\\">
      <div class=\\"el-divider__text is-center\\">
        <div data-v-bd75ece3=\\"\\" class=\\"ion-heading\\">
          <div data-v-bd75ece3=\\"\\" class=\\"ion-link\\">
            <div data-v-bd75ece3=\\"\\">
              <!--v-if--><span data-v-bd75ece3=\\"\\" class=\\"ion-formula\\">[C<sub class=\\"leading-none\\">10</sub>H<sub class=\\"leading-none\\">11</sub>NO + Na]⁺</span>
            </div>
            <div data-v-22a3ab3a=\\"\\" data-v-bd75ece3=\\"\\" class=\\"fdr-badge fdr-badge-10\\">10% FDR</div>
            <!--v-if-->
          </div>
          <!--v-if-->
        </div>
      </div>
    </div>
    <div data-v-57c423f9=\\"\\" data-v-bd75ece3=\\"\\" class=\\"compound-list\\">
      <div data-v-57c423f9=\\"\\" class=\\"compound\\">
        <div data-v-57c423f9=\\"\\" class=\\"mock-el-popover\\" placement=\\"left\\" trigger=\\"click\\">
          <div data-v-57c423f9=\\"\\" class=\\"compound-popover\\">
            <figure data-v-57c423f9=\\"\\" class=\\"flex flex-col items-center justify-center\\">
              <figcaption data-v-57c423f9=\\"\\">Tryptophanol <br data-v-57c423f9=\\"\\"><a data-v-57c423f9=\\"\\" href=\\"http://www.hmdb.ca/metabolites/HMDB0003447\\" target=\\"_blank\\"> View on HMDB-v4 website </a></figcaption><img data-v-57c423f9=\\"\\" src=\\"/mol-images/HMDB/HMDB0003447.svg\\" class=\\"compound-image\\">
            </figure>
          </div>
        </div><br data-v-57c423f9=\\"\\"><span data-v-57c423f9=\\"\\"><a data-v-57c423f9=\\"\\" href=\\"http://www.hmdb.ca/metabolites/HMDB0003447\\" target=\\"_blank\\">Tryptophanol</a></span>
      </div>
      <!--v-if-->
    </div>
  </div>
  <div data-v-bd75ece3=\\"\\">
    <div data-v-bd75ece3=\\"\\" class=\\"el-divider el-divider--horizontal\\" style=\\"--el-border-style: solid;\\" role=\\"separator\\">
      <div class=\\"el-divider__text is-center\\">
        <div data-v-bd75ece3=\\"\\" class=\\"ion-heading\\">
          <div data-v-bd75ece3=\\"\\" class=\\"ion-link\\">
            <div data-v-bd75ece3=\\"\\"><span data-v-bd75ece3=\\"\\">Isomer:</span><span data-v-bd75ece3=\\"\\" class=\\"ion-formula\\">[C<sub class=\\"leading-none\\">10</sub>H<sub class=\\"leading-none\\">13</sub>NO<sub class=\\"leading-none\\">2</sub> + Na - H<sub class=\\"leading-none\\">2</sub>O]⁺</span></div>
            <div data-v-22a3ab3a=\\"\\" data-v-bd75ece3=\\"\\" class=\\"fdr-badge fdr-badge-20\\">20% FDR</div>
            <!--v-if-->
          </div>
          <div data-v-bd75ece3=\\"\\" class=\\"mock-el-popover\\" trigger=\\"hover\\" placement=\\"top\\">
            <div data-v-bd75ece3=\\"\\" style=\\"max-width: 500px;\\">
              <p data-v-bd75ece3=\\"\\"> The False Discovery Rate (FDR) for each annotation is calculated among all ions that share the same adduct. </p>
              <p data-v-bd75ece3=\\"\\"> It is possible for isomeric annotations to show different FDRs due to having different adducts. In these cases, it is an indicator of certain adducts having a higher or lower probability of incorrectly labelling unknown molecules. </p>
              <p data-v-bd75ece3=\\"\\">The FDR should not be used to decide which isomeric molecule is more likely to be correct.</p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div data-v-57c423f9=\\"\\" data-v-bd75ece3=\\"\\" class=\\"compound-list\\">
      <div data-v-57c423f9=\\"\\" class=\\"compound\\">
        <div data-v-57c423f9=\\"\\" class=\\"mock-el-popover\\" placement=\\"left\\" trigger=\\"click\\">
          <div data-v-57c423f9=\\"\\" class=\\"compound-popover\\">
            <figure data-v-57c423f9=\\"\\" class=\\"flex flex-col items-center justify-center\\">
              <figcaption data-v-57c423f9=\\"\\">Maltoxazine <br data-v-57c423f9=\\"\\"><a data-v-57c423f9=\\"\\" href=\\"http://www.hmdb.ca/metabolites/HMDB0030372\\" target=\\"_blank\\"> View on HMDB-v4 website </a></figcaption><img data-v-57c423f9=\\"\\" src=\\"/mol-images/HMDB/HMDB0030372.svg\\" class=\\"compound-image\\">
            </figure>
          </div>
        </div><br data-v-57c423f9=\\"\\"><span data-v-57c423f9=\\"\\"><a data-v-57c423f9=\\"\\" href=\\"http://www.hmdb.ca/metabolites/HMDB0030372\\" target=\\"_blank\\">Maltoxazine</a></span>
      </div>
      <div data-v-57c423f9=\\"\\" class=\\"compound\\">
        <div data-v-57c423f9=\\"\\" class=\\"mock-el-popover\\" placement=\\"left\\" trigger=\\"click\\">
          <div data-v-57c423f9=\\"\\" class=\\"compound-popover\\">
            <figure data-v-57c423f9=\\"\\" class=\\"flex flex-col items-center justify-center\\">
              <figcaption data-v-57c423f9=\\"\\">Salsolinol <br data-v-57c423f9=\\"\\"><a data-v-57c423f9=\\"\\" href=\\"http://www.hmdb.ca/metabolites/HMDB0042012\\" target=\\"_blank\\"> View on HMDB-v4 website </a></figcaption><img data-v-57c423f9=\\"\\" src=\\"/mol-images/HMDB/HMDB0042012.svg\\" class=\\"compound-image\\">
            </figure>
          </div>
        </div><br data-v-57c423f9=\\"\\"><span data-v-57c423f9=\\"\\"><a data-v-57c423f9=\\"\\" href=\\"http://www.hmdb.ca/metabolites/HMDB0042012\\" target=\\"_blank\\">Salsolinol</a></span>
      </div>
      <!--v-if-->
    </div>
  </div>
  <!--v-if-->
</div>"
`;
